packages/core/src/common/error.ts

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+export enum ApplicationErrorKind {
+    FRONTEND = "ERR_THEIA_FRONTEND",
+    BACKEND = "ERR_THEIA_BACKEND"
+}
+
+export class ApplicationError extends Error {
+
+    /**
+     * Static code member representing the error
+     */
+    static code: string = "ERR_THEIA_DEFAULT";
+
+    /**
+     * Instance property representing the error
+     */
+    code: string;
+
+    /**
+     * What kind of error is being thrown (backend or frontend)
+     */
+    readonly kind: ApplicationErrorKind;
+
+    /**
+     * Creates an error and overriding its error code
+     *
+     * @param code error code defining the error
+     * @param message error message detailling the reason
+     */
+    static withCode(code: string, message?: string): ApplicationError {
+        const error = new this(message);
+        error.code = code;
+        return error;
+    }
+
+    constructor(message?: string) {
+        super(message);
+        this.code = new.target.code; // Applying default
+        Object.setPrototypeOf(this, new.target.prototype);
+    }
+
+    toJSON() {
+        return {
+            message: this.message,
+            stack: this.stack,
+            code: this.code,
+            kind: this.kind,
+        };
+    }
+}
+
+/**
+ * Errors from the Frontend
+ */
+export class FrontendError extends ApplicationError {
```

Comment:
we don't need `FrontendError` and `BackendError`,  `ApplicationError` will be enough